File: implement-shell-tools/cat/cat.js
const fs = require("fs");

const args = process.argv.slice(2);

let numberAllLines = false;
let numberNonBlankLines = false;
let startIndex = 0;
let lineNumber = 1;

// Check for flags
for (let i = 0; i < args.length; i++) {
  if (args[i] === "-n") {
    numberAllLines = true;
    startIndex++;
  } else if (args[i] === "-b") {
    numberNonBlankLines = true;
    startIndex++;
  } else {
    break;
  }
}

// Read each file
for (let i = startIndex; i < args.length; i++) {
  let fileName = args[i];

  try {
    let content = fs.readFileSync(fileName, "utf8");

    // Remove the extra empty line if the file ends with "\n"
    let lines = content.endsWith("\n")
      ? content.slice(0, -1).split("\n")
      : content.split("\n");

    for (let j = 0; j < lines.length; j++) {
      let line = lines[j];

      if (numberAllLines) {
        console.log(lineNumber + " " + line);
        lineNumber++;
      } else if (numberNonBlankLines) {
        if (line.trim() === "") {
          console.log(line);
        } else {
          console.log(lineNumber + " " + line);
          lineNumber++;
        }
      } else {
        console.log(line);
      }
    }
  } catch (error) {
    console.log("Cannot read file: " + fileName);
  }
}

File: implement-shell-tools/cat/cat.py
import sys

# Check if the user used a flag
if len(sys.argv) > 1 and sys.argv[1] == "-n":

    # Start from the second argument because the first is "-n"
    for filename in sys.argv[2:]:

        with open(filename, "r") as file:

            line_number = 1

            for line in file:
                print(f"{line_number} {line}", end="")
                line_number += 1


elif len(sys.argv) > 1 and sys.argv[1] == "-b":

    # Start from the second argument because the first is "-b"
    for filename in sys.argv[2:]:

        with open(filename, "r") as file:

            line_number = 1

            for line in file:

                # Only number non-empty lines
                if line.strip() == "":
                    print(line, end="")
                else:
                    print(f"{line_number} {line}", end="")
                    line_number += 1


else:

    # No flag, just print every file
    for filename in sys.argv[1:]:

        with open(filename, "r") as file:
            print(file.read(), end="")